# 7. Theoretical question

#### 1. Try solving the problem mentioned above using three different sorting algorithms (do not use any MapReduce algorithm). (Note: Built-in Python functions (like .mean, .sort, etc.) are not allowed to be used. You must implement the algorithms from scratch).

In [2]:
#let's import the only library we'll need that is time and we'll use it to evaluate the time taken by the algorithm
import time

In the next blocks there are some useful functions that we'll use for answering question 7.1:
- **my_mean** : it is a function that given a row of the file ApplicantsInfo.txt returns the average of the m scores in that row
- **sort_by_name** : it is a function that, given two words (name1, name2), if name2 in alphabetical order is before name1 it returns 1, otherwise 0

In [1]:
def my_mean(l,m): 
    mean=0
    for i in range(2,m):
        mean += int(l[i])
    mean = mean/m
    return mean

In [3]:
def sort_by_name(name2,name1):

    for i in range(min(len(name1),len(name2))):
        # for loop stops as soon as the first word finishes the characters
        if name2[i]<name1[i]:
            return 1
        elif name2[i]>name1[i]:
            return 0

    # at this point if the for loop eneded without returning any value this mean that the shortest word is equal to the first part of 
    # the other word and so we'll return 0 or 1 depending on the lengths of the words 
    if len(name1)<len(name2):
        return 0
    else:
        return 1

The next block contains the code to read the file ApplicantsInfo.txt.   
In the first row there are n (number of students), m (number of scores for each student).    
For each of the following n rows there are firts name, second name and all the marks of each student.

In [13]:
f = open("C://Users//auror//Desktop//data_science//ADM//HM//3//ApplicantsInfo.txt","r")
dati=[]

# first we read the first row and we store in n the number of students and in m the number of scores for each student
n, m = map(int,f.readline().split()) 

#starting from the second row, with the while loop we read 1 row at time and the while loop ends when there isn't anymore rows to be read 
while True:
    # dati_temp will be a list s.t. in the first two positions there are 'name' and 'surname' of a student and starting from position 3 till
    # the last we have all of his/her scores (m in total)
    dati_temp = f.readline().split()
    if not dati_temp:
        break
    # then we call the function my_mean in order to compute the mean of the scores of a given student  and then we'll append to the list 'dati'
    # a list done like this [name,surname,mean of the scores]
    mean = my_mean(dati_temp,m)
    dati.append([dati_temp[0],dati_temp[1],round(mean,2)])
f.close()

The next 3 blocks contains the code to compute each of the sorting algorithm.   
**NOTE**: if you execute one of the three following cell and then you wanna compute another sorting algorithm you should before rerun the block in which we store the data in the list dati.

In [15]:
#run this cell tu compute selection sort algorithm
start_time = time.time()
insertion_sort(dati,n)
end_time = time.time()
minuti = (end_time-start_time)//60
secondi = (end_time-start_time)%60
print('The time taken for computing insertion_sort is: ',minuti,'minutes and',round(secondi,2),'seconds')

The time taken for computing insertion_sort is:  6.0 minutes and 35.15 seconds


In [27]:
##run this cell tu compute merge sort algorithm
start_time = time.time()
merge_sort(1,n)
end_time = time.time()
minuti = (end_time-start_time)//60
secondi = (end_time-start_time)%60
print('The time taken for computing insertion_sort is: ',minuti,'minutes and',round(secondi,2),'seconds')

The time taken for computing insertion_sort is:  0.0 minutes and 0.89 seconds


The next block contains the code to write a .text file with the final output

In [16]:
f= open("RankingList_insertion_sort.txt","w")
for i in dati:
    f.write(i[0]+" "+i[1]+" "+str(i[2]))
    f.write("\n")
f.close()

The next block contains the function insertion_sort

In [14]:
#selection sort
def insertion_sort(dati,n):
    # the for loop starts from the second element of the list dati since we suppose that the first element is already in the right position
    for i in range(1,int(n)):
        # starting from the element i-1 we go backwords in the list as soon as we found an element j such that dati[j][2]<dati[i][2] or 
        # dati[j][2]=dati[i][2] but in alphabetical order the name and surname of the element j have to be before the element i 
        while ((dati[i][2]>=dati[i-1][2]) & (i>0)):
            # everytime if dati[i][2]>dati[i-1][2] we exchange the two lists (dati[i] and dati[i-1])
            if dati[i][2]>dati[i-1][2]:
                temp=dati[i-1]
                dati[i-1]=dati[i]
                dati[i]=temp
                # after exchanging element i and element i-1 we point to the element before i and skip the rest of the while since we know that 
                # the avarages were not equal
                i -= 1
                continue
            # if dati[i][2]=dati[i-1][2] then we check the alphabetical order of the two lists
            elif dati[i][2]==dati[i-1][2]:
                # if the names of element i and element i-1 are equal we call the function sort_by_name with their surnames
                if dati[i][0]==dati[i-1][0]:
                    y_n = sort_by_name(dati[i][1],dati[i-1][1])
                #instead if the names are not equal we call the function sort_by_name with the names of element i and i-1
                elif dati[i][0]!=dati[i-1][0]:
                    y_n = sort_by_name(dati[i][0],dati[i-1][0])
                # if y_n=1 this mean that name and surname of dati[i] in alphabetical order have to be before dati[i-1] and so we exchange the 
                # two lists 
                if y_n:
                    temp=dati[i-1]
                    dati[i-1]=dati[i]
                    dati[i]=temp
                    i -= 1
                # otherwise, if element i and i-1 were already in alphabetical order this mean that the element i was already in the right
                # position and so we break the while 
                else:
                    break
            
            #i -= 1
    return

The next two blocks contains the merge_sort function and the merge function used in the first one.

In [12]:
def merge_sort(start,end):
    if start<end:
        medium = (start+end)//2
        merge_sort(start,medium)
        merge_sort(medium+1,end)
        merge(start,medium,end)
    return

In [13]:
def merge(start,medium,end):
    B = []
    i = start-1
    j = medium
    k = 0
    
    while ((i<medium) and (j<end)):
        if dati[i][2]<dati[j][2]:
            B.append(dati[j])
            j+=1
        elif dati[i][2]>dati[j][2]:
            B.append(dati[i])
            i+=1
        elif dati[i][2]==dati[j][2]:
            if dati[i][0]==dati[j][0]:
                y_n = sort_by_name(dati[i][1],dati[j][1])
            elif dati[i][0]!=dati[j][0]:
                y_n = sort_by_name(dati[i][0],dati[j][0])
            if y_n==1:
                B.append(dati[i])
                i+=1
            elif y_n==0:
                B.append(dati[j])
                j+=1
        k+=1
    
    while i<medium:
        B.append(dati[i])
        i+=1
        k+=1
    while j<end:
        B.append(dati[j])
        j+=1
        k+=1
    
    if B !=[]:
        for l in range(0,end-start+1):
            dati[start-1+l]=B[l]
    return 

**solo una prova per vedere se gli output dei vari algoritmi coincisono**

In [17]:
f1 = open("RankingList_merge_sort.txt","r")
f2 = open("RankingList_insertion_sort.txt","r")

for i in range(1,n+1):
    if f1.readline() != f2.readline():
        print('la vita fa schifo')
        break
    
    
print('la vita è bella',i)
f1.close()
f2.close()

la vita è bella 50000


#### 2. What is the time complexity of each algorithm you have used?

#### 3. Evaluate the time taken for each of your implementations to answer the query stored in the ApplicantsInfo.txt file and visualize them.

As we've seen before the times taken by the different algorithm are:
- **insertion sort**: 6.0 minutes and 35.15 seconds
- **merge_sort**: 0.0 minutes and 0.89 seconds

#### 4. What is the most optimal algorithm, in your opinion, and why?

#### 5. Implement a sorting algorithm using MapReduce and compare it against the three algorithms previously implemented using the ApplicantsInfo.txt file.